File: app/src/main/java/com/nicholasworkshop/tinklabstest/external/content/ContentServiceManager.java
package com.nicholasworkshop.tinklabstest.external.content;

import com.nicholasworkshop.tinklabstest.external.content.model.Story;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by nicholas.wong on 2017/05/15.
 */
@Singleton
public class ContentServiceManager {

    private final ContentService contentService;

    @Inject
    ContentServiceManager(ContentService contentService) {
        this.contentService = contentService;
    }

    public Observable<List<Story>> getCityGuide(int count) {
        return contentService
                .getCityGuide()
                .cache()
                .map(stories -> stories.subList(0, Math.min(stories.size(), count)))
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Observable<List<Story>> getEatGuide(int count) {
        return contentService
                .getEatGuide()
                .cache()
                .map(stories -> stories.subList(0, Math.min(stories.size(), count)))
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Observable<List<Story>> getShopGuide(int count) {
        return contentService
                .getShopGuide()
                .cache()
                .map(stories -> stories.subList(0, Math.min(stories.size(), count)))
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }
}

File: app/src/main/kotlin/com/nicholasworkshop/tinklabstest/external/content/ContentServiceManager.kt
package com.nicholasworkshop.tinklabstest.external.content

import com.nicholasworkshop.tinklabstest.external.content.model.Story
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ContentServiceManager
@Inject
internal constructor(private val contentService: ContentService) {

    fun getCityGuide(count: Int): Observable<List<Story>> {
        return contentService
                .cityGuide()
                .cache()
                .map { it.subList(0, Math.min(it.size, count)) }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
    }

    fun getEatGuide(count: Int): Observable<List<Story>> {
        return contentService
                .eatGuide()
                .cache()
                .map { it.subList(0, Math.min(it.size, count)) }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
    }

    fun getShopGuide(count: Int): Observable<List<Story>> {
        return contentService
                .shopGuide()
                .cache()
                .map { it.subList(0, Math.min(it.size, count)) }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
    }
}
